heat/core/linalg.py

```diff
@@ -1,15 +1,459 @@
+
+        elif a.split is None and b.split == 0:
+            c = torch.zeros((a.gshape[-2], b.gshape[1]))
+            b_idx = b.comm.chunk(b.shape, b.split)[2]
+            c += a._DNDarray__array[:, b_idx[0].start:b_idx[0].start + b.lshape[0]] @ b._DNDarray__array
+            b.comm.Allreduce(MPI.IN_PLACE, c, MPI.SUM)
+            return factories.array(c, split=b.split)
+
+        elif a.split == 0 and b.split == 0:
+            split_0_flag = True
+        elif a.split == 1 and b.split == 1:
+            split_1_flag = True
+        elif a.split == 0 and b.split == 1:
+            split_01_flag = True
+        elif a.split == 1 and b.split == 0:
+            split_10_flag = True
+        else:
+            raise NotImplementedError("splits > 1 not implemented")
+
+        # block sizes dont need to be the same. thy just need the same inner dimmension (kB)
+        kB = 0
+        rem_a, rem_b = [0] * 2
+        if a.split == len(a.gshape)-1 and b.split == len(a.gshape)-2:  # if the split direction is the last dim in a and the first dim in b
+            # the max inner dim (kB) is the min value from the result of the integer division of the last dim of a/world size and the first dim of b/world size
+            kB = min([a.gshape[-1] // a.comm.size, b.gshape[0] // b.comm.size])
+        elif a.split == len(a.gshape) - 2 and b.split == len(a.gshape) - 1:
+            kB = a.gshape[-1]
+        elif a.split == len(a.gshape)-1:
+            kB = a.gshape[-1] // a.comm.size
+        elif b.split == len(a.gshape)-2:
+            kB = b.gshape[0] // b.comm.size
+            kB = kB if kB < a.gshape[-1] else a.gshape[-1]
+
+        if a.lshape[-1] % kB != 0:
+            rem_a = 1
+        if b.lshape[0] % kB != 0:
+            rem_b = 1
+
+        # get the lshape map to determine what needs to be sent where as well as M and N
+        # lshape map dims -> {node, a=0, b=1, lshape}
+        lshape_map = factories.zeros((a.comm.size, 2, len(a.gshape)), dtype=int)
+        lshape_map[a.comm.rank, 0, :] = torch.Tensor(a.lshape)
+        lshape_map[b.comm.rank, 1, :] = torch.Tensor(b.lshape)
+        a.comm.Allreduce(MPI.IN_PLACE, lshape_map, MPI.SUM)
+
+        # find mB (first blocking dim for a) and nB (2nd blocking dim for b)
+        mB = lshape_map[:, 0, -2].min().item()
+        nB = lshape_map[:, 1, -1].min().item()
+
+        # check for remaining dims in the outside dimensions
+        rem_a_out, rem_b_out = 0, 0
+        if a.lshape[-2] % mB != 0:
+            rem_a_out = 1
+        if b.lshape[-1] % nB != 0:
+            rem_b_out = 1
+
+        # get the flags from all processes
+        # rem_map dims guide -> {process number, a/b (0/1), True/False (1/0) if there is a remainder in this dimension
+        rem_map = factories.zeros((a.comm.size, 2, 2))
+        rem_map[a.comm.rank, 0, :] = (rem_a_out, rem_a)
+        rem_map[a.comm.rank, 1, :] = (rem_b, rem_b_out)
+        a.comm.Allreduce(MPI.IN_PLACE, rem_map, MPI.SUM)
```

Comment:
Can we do an Asynchronous allreduce here, i.e. Iallreduce() and wait for the request later on? Same goes for l. 176 an so one. This way we could overlap communication and computation slightly.

Reply:
The wait would happen on line 188?!